File: 1001-2000/1248. Count Number of Nice Subarrays.java
// Medium
// PrefixSum
// Time:O(n),Space:O(n)
// https://leetcode.cn/problems/count-number-of-nice-subarrays/


class Solution {
  public int numberOfSubarrays(int[] nums, int k) {
      int n = nums.length;
      int[] prefixSum = new int[n + 1];
      prefixSum[0] = 1;

      int cnt = 0, res = 0;
      for (int i = 0; i < n; i++) {
          cnt += nums[i] % 2;
          if (cnt >= k) {
              res += prefixSum[cnt - k];
          }
          prefixSum[cnt]++;
      }
      return res;
  }
}

File: 1001-2000/1248. Count Number of Nice Subarrays.py
# Time:O(n), Space:O(n)

from collections import defaultdict
from typing import List

class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        # prefix_odd_cnt[j] - prefix_even_cnt[i] = k
        # => prefix_even_cnt[i] = prefix_odd_cnt[j] - k
        count = defaultdict(int)
        count[0] = 1

        cur = 0
        res = 0

        for num in nums:
            # odd +1, even +0
            cur += num % 2
            res += count[cur - k]
            count[cur] += 1

        return res
